File: interviews/leetcode/binary-tree-dfs/1448. Count Good Nodes in Binary Tree/solution_iterative.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if not root:
            return 0

        # Stack to store nodes and the max value encountered on the path
        stack = [(root, root.val)]
        count = 0

        while stack:
            node, max_val = stack.pop()

            # If the current node is a good node, increment the count
            if node.val >= max_val:
                count += 1

            # Update the max value for the path and add the children to the stack
            if node.left:
                stack.append((node.left, max(max_val, node.val)))
            if node.right:
                stack.append((node.right, max(max_val, node.val)))

        return count

# Helper function to create a binary tree from a list (for testing purposes)
def create_binary_tree(lst):
    if not lst:
        return None
    root = TreeNode(lst[0])
    queue = [root]
    i = 1
    while i < len(lst):
        node = queue.pop(0)
        if lst[i] is not None:
            node.left = TreeNode(lst[i])
            queue.append(node.left)
        i += 1
        if i < len(lst) and lst[i] is not None:
            node.right = TreeNode(lst[i])
            queue.append(node.right)
        i += 1
    return root

# Test the solution
# if __name__ == "__main__":
#     solution = Solution()
#     root = create_binary_tree([3, 1, 4, 3, None, 1, 5])
#     result = solution.goodNodes(root)
#     print(result)  # Output: 4

File: interviews/leetcode/leetcode-75/binary-tree-dfs/1448. Count Good Nodes in Binary Tree/solution_iterative.go
package main

import "fmt"

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func goodNodes(root *TreeNode) int {
	if root == nil {
		return 0
	}

	// Stack to store nodes and the max value encountered on the path
	stack := []struct {
		node   *TreeNode
		maxVal int
	}{
		{root, root.Val},
	}

	count := 0
	for len(stack) > 0 {
		current := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		node, maxVal := current.node, current.maxVal

		// If the current node is a good node, increment the count
		if node.Val >= maxVal {
			count++
		}

		// Update the max value and add children to the stack
		if node.Left != nil {
			stack = append(stack, struct {
				node   *TreeNode
				maxVal int
			}{node.Left, max(maxVal, node.Val)})
		}
		if node.Right != nil {
			stack = append(stack, struct {
				node   *TreeNode
				maxVal int
			}{node.Right, max(maxVal, node.Val)})
		}
	}

	return count
}

// Helper function to create a binary tree from a slice
func createBinaryTree(lst []interface{}) *TreeNode {
	if len(lst) == 0 || lst[0] == nil {
		return nil
	}
	root := &TreeNode{Val: lst[0].(int)}
	queue := []*TreeNode{root}
	i := 1
	for i < len(lst) {
		node := queue[0]
		queue = queue[1:]
		if lst[i] != nil {
			node.Left = &TreeNode{Val: lst[i].(int)}
			queue = append(queue, node.Left)
		}
		i++
		if i < len(lst) && lst[i] != nil {
			node.Right = &TreeNode{Val: lst[i].(int)}
			queue = append(queue, node.Right)
		}
		i++
	}
	return root
}

func main() {
	root := createBinaryTree([]interface{}{3, 1, 4, 3, nil, 1, 5})
	result := goodNodes(root)
	fmt.Println(result) // Output: 4
}
